# Chapter 1

## 오류의 두 가지 종류
1. 컴파일 오류

    컴파일 단계에서 오류 발생(코드를 잘못 씀)

2. 런타임 오류

    실행 단계에서 오류 발생(DB에 데이터가 없음, 0으로 나눔 등)

## Print
```python
print("Hello python", 123, end="!!!!!")
# 매개변수에 *values: object 이므로 튜플을 받는다.
# 그러므로 print는 여러개의 argument를 튜플의 형태로 받는다.
# end = "\n" 였는데 \n 대신에 !!!!!로 바꿨기 때문에 개행 하지 않는다

print("진짜로 개행 안 하는지 확인") # ㅇㅇ 개행 안 했네

# print(end="!!!", values=("Hello python", "test")) 이런 개념도 있기는 하다
```

기본적으로 여러개를 받을 수 있다

## Type

기본 타입 중 string 타입에는 여러 함수를 적용할 수 있다
```python
number = 10
name = "강수림수림" # 다른 건 그냥 그런데 얘가 좀 중요 : 여러 함수 사용 가능
isOpen = True

print(number)
# print(name_) # 오류 앞 라인까지는 실행 된다
print(name)
print(isOpen)

# print(name.index("숭"))  # exception
# print(name.find("숭"))   # return -1

print(name.index("수",2))
print(name.find("수"))

print("이름은: {name}, 숫자: {number}, 열림: {isOpen}".format(name=name, number=number, isOpen=isOpen))
print(f"이름은: {name}, 숫자: {number}, 열림: {isOpen}") # f 포매팅 : 정처기에 나온 적 있음, 여기서는 변수를 넣었지만 함수 등도 사용 가능
address = """부
산
동
래
구
중앙대로""" # 여러 줄로 나누어 쓸 수 있음
print(address)


```

## Collection

### 컬렉션의 종류

1. set
2. list
3. dict
4. tuple

### 실행 예
```python
set1 = set()
list1 = list()
dict1 = dict()
tuple1 = tuple()

print(type(set1)) # 자료형을 확인하는 type 함수

list2 = []
dict2 = {}
dict3 = {
    "n":0
}
tuple2 = ()
tuple3 = (10) # 10 <- 값이 하나라서 ()를 연산 기호로 본다
tuple4 = (10, ) # ,가 들어가면 그 때 부터 튜플로 본다
set2 = {0}


print(list2) # []
print(dict2) # {}
print(tuple2) # ()
print(set2) # {0}
print(dict3) # {'n': 0}

print(type(()))
```

## List

### 슬라이싱
list의 index를 나타내는 [] 내에 `:`을 사용하여 특정 index의 값(value)을 반환하는 것
```python
numbers = [1,2,3,4,5]

print(numbers[0])   # index
print(numbers[0:3]) # index to index
print(numbers[:3])  # start to end
print(numbers[3:])  # index to end
print(numbers[-3:]) # -index to end
print(numbers[:-2]) # start to -index
print("부산 동래구"[3:]) # string에서도 마찬가지로 index to end
print((1,2,3,4)[2:])    # 튜플도 마찬가지
# print({1,2,3,4}[2:])  # set은 순서가 없어서 안 된다
print([1,2,3,4][2:])    # 마찬가지로 list도 생성과 동시에 사용 가능

# 원하는 내용 출력하기
anyList = [[1, "김준일", [3, 4, [7, "김준이"], 6]]]

# 1. 김준이를 출력하시오
print(anyList[0][2][2][1])

# 2. [3, 4, [7, '김준이'], 6]를 출력하시오
print(anyList[0][2][:])

# 3. [4, [7, '김준이']]를 출력하시오
print(anyList[0][2][2:3])   # index(start) to index(end)
```

## 개인 추가 학습

### for문을 사용한 다차원 리스트 value 추출하기

```python
listList = [[[0,1,2], [3,4,5], [6,7,8]],
            [[9,10,11],[12,13,14],[15,16,17]],
            [[18,19,20],[21,22,23],[24,25,26]]]

listK = []
n = 0

for i in listList:
    for j in i:
        for k in j: # 여기에서 k는 각각의 원자값(리스트가 아닌 그 value)을 가진다
            listK.append(k)
            print(listK[n], end=" ")
            n += 1
print()
# 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 

print(listK[:])
# [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
```

## 연산자

```python
#연산자
"""
+, -, *, /, %
"""

print(7/4) # 1.75
print(7//4) # 1 <- 몫만 가져옴
print(7%4) # 3 <- 이건 나머지
print(7**4) # 2401 <- 7^4
print()

print("*", "\n", "*" * 2, "\n", "*" * 3, sep="") # 문자열에 곱하기 n을 사용하면 해당 문자열을 n번 반복하여 반환한다
```

## 반복문

### while
   **기본 형태**
```python
i = 0

while i < 5:
    print("내용 작성")
```

### while을 통한 list 출력하기

```python
anyList = [1,2,3,4]
i = 0
while i < len(anyList):
    print(anyList[i])
    i += 1
```

## list 관련 함수

### len()

```python
print(len([1,2,3])) # len() 함수
                    # list.len X / len(list)
print()
```

### 자료형 : int(), str(), bool()
```python
print(str(3) + "hi")    # 3hi
print(int("1") + 2)     # 3
                        # 자료형 함수를 사용하면 casting이 가능하다
print(bool("True") & False) # False
```

### list 삭제 : del, remove(), pop()
```python
anyList = [1,2,3,4]

del anyList[2]

print(anyList)

# del anyList[4] # del 키워드로 삭제할 때는 값이 꼭 존재해야

anyList.remove(1)
print(anyList)

anyList.pop(1)  # pop() 이렇게 아무것도 넣지 않으면 마지막 index의 값이 튀어나오고 삭제된다.
print(anyList)


try:
    anyList.pop(4)
    print(anyList)
except Exception as e:
    print("해당 리스트의 범위를 초과한 참조이다.")
```

- 예외 처리를 위해 가능하다면 함수를 사용하도록 하자 (del 지양)
- .remove()와 .pop()은 앞의 내용이 list인지 1차적으로 확인할 수 있고
- 예외 처리를 위한 try - except 문에 사용할 수 있다

   `Except`에 `기분 나쁜 노란 줄`이 생긴 이유는 해당 코드가 `검출 가능한 오류의 범위`가 try 내에 작성한 오류 보다 `너무 크기 때문`이다. 여기서는 IndexError로 수정하면 기분 나쁜 노란 줄이 사라지지만 일단 놔둠.